Client sampling and roots capabilities set to None if not implemented

lorenzocesconetto opened the pull request:
Right now clients always claims to support sampling and roots capabilities, which is inaccurate.
Having this properly set allows servers to fallback to alternative implementations.

## Motivation and Context
Servers might rely on the correctness of client capabilities, so clients should inform them accurately.

## How Has This Been Tested?
Unit tests

## Breaking Changes
No

## Types of changes
<!-- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation update

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have read the [MCP Documentation](https://modelcontextprotocol.io)
- [x] My code follows the repository's style guidelines
- [x] New and existing tests pass locally
- [x] I have added appropriate error handling
- [x] I have added or updated documentation as needed

## Additional context
<!-- Add any other context, implementation notes, or design decisions -->

lorenzocesconetto commented:
@ihrpr I think this is an important PR to make sure the SDK adheres to the protocol.
This enables servers to make decisions based on the client capabilities.
For instance, server won't waste time trying to send a sampling request to a client that has no sampling capabilities.

hesreallyhim commented on src/mcp/client/session.py:
```diff
@@ -116,13 +116,20 @@ def __init__(
+            roots = None
+        else:
+            roots = types.RootsCapability(
+                # TODO: Should this be based on whether we
```
The spec (says)[https://modelcontextprotocol.io/specification/2025-03-26/client/roots#root-list-changes]:

> When roots change, clients that support listChanged MUST send a notification:

So I guess it would be the former.

lorenzocesconetto replied:
Totally agree!
This `TODO` comment was already in the code base, I just kept it...
I only added the `if-else` structure around it.

ihrpr commented on src/mcp/client/session.py:
```diff
@@ -116,13 +116,20 @@ def __init__(
+            if self._sampling_callback is not _default_sampling_callback
+            else None
         )
+        if self._list_roots_callback is _default_list_roots_callback:
```
nit: would be good to match the style of sampling here

ihrpr commented on src/mcp/client/session.py:
```diff
@@ -116,13 +116,20 @@ def __init__(
+            roots = None
+        else:
+            roots = types.RootsCapability(
+                # TODO: Should this be based on whether we
```
let's remove TODO

ihrpr reviewed (approved):
Thank you! 

Please can can those nits be addressed and we can merge.

lorenzocesconetto commented:
@ihrpr I've addressed your comments, but since there's a new commit, we need your approval again.

Btw, I want to implement a larger feature, but before actually implementing it and opening a PR, I'd love to chat with you just to make sure it makes sense. Is there a way I can reach out to you @ihrpr ?

ihrpr commented on src/mcp/client/session.py:
```diff
@@ -116,12 +116,15 @@ def __init__(
+            else None
+        )
+        roots = (
+            types.RootsCapability(listChanged=True)
```
Maybe worth putting the comment back, sorry.

The client _can_ send list changes, as the code is there to trigger that notification, so it's supported.
But the spec says "listChanged indicates whether the client will emit notifications when the list of roots changes." 
We do need to address this, but a separate PR would be probably be better

lorenzocesconetto commented:
@ihrpr I want to implement a pretty large feature, but before actually implementing it and opening a PR, I'd love to chat with you. Is there a way I can reach out to you?

ihrpr commented:
> @ihrpr I want to implement a pretty large feature, but before actually implementing it and opening a PR, I'd love to chat with you. Is there a way I can reach out to you?

oh, sorry, missed this. I think the best option would be to open an issue and we can discuss there

lorenzocesconetto commented:
Sounds good! @ihpr

Btw I am working on a fix for the `rootListChanged`. Thank you @hesreallyhim for pointing that out.

hesreallyhim commented:
> Sounds good! @ihpr
> 
> Btw I am working on a fix for the `rootListChanged`. Thank you @hesreallyhim for pointing that out.

thanks @ihrpr ! hope I haven't added to the confusion, please see my last comment/essay that I posted for a thorough recap of my position

lorenzocesconetto commented:
@hesreallyhim imo your understanding is right. The client capabilities declaration was obviously not following the protocol specs, that's why I opened this PR in the first place. This PR addressed only part of the issue, as you've noted the hardcoded piece `listChanged=True` in the Roots capability still needs to be fixed. I am working on a fix and will open a new PR to address it.
